wgengine/magicsock/magicsock.go

```diff
@@ -2393,6 +2396,42 @@ func (c *Conn) closeAllDerpLocked(why string) {
 	c.logActiveDerpLocked()
 }
 
+// maybeCloseDerpOnRebindLocked, in response to a rebind, closes all
+// DERP connections that don't have a local address in okayLocalIPs
+// and pings all those that do.
+//
+// c.mu must be held.
+func (c *Conn) maybeCloseDerpOnRebindLocked(okayLocalIPs []netaddr.IPPrefix) {
+	for regionID, ad := range c.activeDerp {
+		la, err := ad.c.LocalAddr()
+		if err != nil {
+			c.closeDerpLocked(regionID, "rebind-no-localaddr")
+			continue
+		}
+		if !tsaddr.PrefixesContainsIP(okayLocalIPs, la.IP()) {
+			c.closeDerpLocked(regionID, "rebind-default-route-change")
+			continue
+		}
+		regionID := regionID
+		dc := ad.c
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+			defer cancel()
+			if err := dc.Ping(ctx); err != nil {
+				c.mu.Lock()
+				defer c.mu.Unlock()
+				c.closeDerpLocked(regionID, "rebind-ping-fail")
+				if !c.privateKey.IsZero() {
+					c.startDerpHomeConnectLocked()
```

Comment:
I don’t understand why this is done only on ping failure

Reply:
That is, don’t we need this on the local IP check above?

Reply:
The caller, Conn.Rebind, does it.

But I can pull it out into a new method like "closeOrReconnectDerpLocked" that makes it more clear.

Reply:
Ah I see now. Maybe move that startDerpHomeConnectLocked from Rebind into this function?